### PR TITLE
Seq bytes

### PR DESCRIPTION
**Pull request by mdehoon**
Currently, `Seq` methods call `bytes` on `self._data` to convert the sequence data to a `bytes` object.
However, in most cases `self._data` already is a `bytes` object (exceptions are when the `Seq` is created by BioSQL or from a twoBit file, or when the sequence contents is unknown). This PR moves the call to `bytes` from the `Seq` methods to the `SequenceDataAbstractBaseClass` base class. This avoids calling `bytes` unnecessarily, but perhaps more importantly, this makes the implementation of the following methods identical for the `Seq` and `MutableSeq` classes:

 - `__str__`
 - `__eq__`
 - `__lt__`
 - `__le__`
 - `__gt__`
 - `__ge__`
 - `__len__`
 - `__add__`
 - `__radd__`
 - `__mul__`
 - `__rmul__`
 - `__imul__`
 - `count`
 - `count_overlap`
 - `__contains__`
 - `find`
 - `rfind`
 - `index`
 - `rindex`
 - `startswith`
 - `endswith`
 - `translate`

To ensure that `Seq` and `MutableSeq` will remain consistent, and to avoid having the exact same implementation twice in `Bio/Seq.py`, we could define an abstract base class that implements these methods. The current PR does not do that yet though.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/Seq.py [hunk 1]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
     def __getitem__(self, key):
         pass
 
+    def __bytes__(self):
+        return self[:]
```
Comment (peterjc) [inline Bio/Seq.py:94]: Should the docstring suggest subclassing ``__bytes__`` too if needed?

File: Bio/Seq.py [hunk 2]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
...
+    def __radd__(self, other):
+        return other + bytes(self)
+
+    def __mul__(self, other):
+        return bytes(self) * other
+
+    def __contains__(self, item):
+        return bytes(self).__contains__(item)
+
+    def decode(self, encoding="utf-8"):
```
Comment (peterjc) [inline Bio/Seq.py:126]: I thought we were assuming ASCII only? I would think it better to start off strict, and relax this later if needed

File: Bio/Seq.py [hunk 3]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
...
+        """Decode the data as bytes using the codec registered for encoding.
+
+        encoding
+          The encoding with which to decode the bytes.
```
Comment (peterjc) [inline Bio/Seq.py:130]: Does this render nicely as RST?

File: Bio/Seq.py [hunk 4]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
...
+        maxsplit
+          Maximum number of splits to do.
+          -1 (the default value) means no limit.
+        """
+        return bytes(self).split(sep, maxsplit)
+
+    def rsplit(self, sep=None, maxsplit=-1):
+        """Return a list of the sections in the data, using sep as the delimiter.
+
+          sep
```
Comment (peterjc) [inline Bio/Seq.py:219]: Over indented?

File: Bio/Seq.py [hunk 5]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
...
+        return bytes(self).upper()
+
+    def lower(self):
+        """Return a copy of data with all ASCII characters converted to lowercase."""
+        return bytes(self).lower()
+
+    def replace(self, old, new):
+        """Return a copy with all occurrences of substring old replaced by new.
+
+          count
```
Comment (peterjc) [inline Bio/Seq.py:263]: There is no count argument?

File: Bio/Seq.py [hunk 6]
```diff
@@ -90,6 +90,196 @@ def __len__(self):
...
+        If the optional argument count is given, only the first count occurrences are
+        replaced.
+        """
+        return bytes(self).replace(old, new)
+
+    def translate(self, table, delete=b""):
+        """Return a copy with each character mapped by the given translation table.
+
+          table
+            Translation table, which must be a bytes object of length 256.
```
Comment (peterjc) [inline Bio/Seq.py:276]: Python 3 is more flexible here, see also #2968

File: Bio/Seq.py [hunk 7]
```diff
@@ -1390,6 +1584,10 @@ def __bytes__(self):
         """Return the unknown sequence as full string of the given length."""
         return self._character.encode("ASCII") * self._length
 
+    @property
+    def _data(self):
+        return self._character.encode("ASCII") * self._length
+
```
Comment (peterjc) [inline Bio/Seq.py:1590]: Defining ``._data`` as a property is rather a neat approach. Nice.

File: Bio/Seq.py [hunk 8]
```diff
@@ -3049,10 +3252,15 @@ def __bytes__(self):
             return b""
         raise UndefinedSequenceError("Sequence content is undefined")
 
+    def __add__(self, other):
+        if isinstance(other, _UndefinedSequenceData):
+            return _UndefinedSequenceData(self._length + other._length)
+        raise TypeError
+
```
Comment (peterjc) [inline Bio/Seq.py:3259]: This breaking of the add method means I can think of a clearly different use case for the now deprecated ``UnknownSeq``, things like ``UnknownSeq(100, character="N")`` versus ``Seq("N"*100)`` for gaps when joining contigs together. The workaround here is obvious though - make the short spacer sequence explicit.

File: Bio/Seq.py (hunk 2 above)
Comment (mdehoon) [reply Bio/Seq.py:126]: I am using "utf-8" here to be consistent with `bytes`.

File: Bio/Seq.py (hunk 6 above)
Comment (mdehoon) [reply Bio/Seq.py:276]: Not for `bytes`:
```python
>>> b"XYZ".translate(b"ABCD")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: translation table must be 256 characters long
```

Comment (peterjc) [reply Bio/Seq.py:276]: Good justification, thank you.

File: Bio/Seq.py (hunk 2 above)
Comment (peterjc) [reply Bio/Seq.py:126]: OK

File: Bio/Seq.py (hunk 5 above)
Comment (mdehoon) [reply Bio/Seq.py:263]: Fixed, thanks. There is no `count` argument because we never actually use it.

File: Bio/Seq.py (hunk 4 above)
Comment (mdehoon) [reply Bio/Seq.py:219]: Fixed

File: Bio/Seq.py (hunk 1 above)
Comment (mdehoon) [reply Bio/Seq.py:94]: The `_BioSQLSequenceData` class in BioSQL indeed was subclassing `__bytes__`, but it is not actually needed if `__getitem__` is implemented appropriately. There may still be some cases where `__bytes__` needs to be subclassed, but by default I would not encourage it if it is not needed.

Comment (peterjc) [reply Bio/Seq.py:104]: Sorry, I meant to ask if the documentation should talk about re-implementing the ``__bytes__`` method? It looks like this is covered by your description of what ``[:]`` should do.

**Comment by mdehoon** [comment]
@peterjc  Good to go?

**Review by peterjc — APPROVED** [review]

**Comment by mdehoon** [comment]
@peterjc  Thank you for looking at this in detail. How do you think about giving `Seq` and `MutableSeq` an abstract base class? It would let us get rid of all the duplicated code that is now in `Bio/Seq.py`.

**Comment by peterjc** [comment]
Adding an ABC sounds good in principle, and it might perhaps be best a private class?

I am assuming we don't want to try making ``MutableSeq`` a subclass of ``Seq`` (for backward compatibility if not ease of implementation).

**Comment by mdehoon** [comment]
> I am assuming we don't want to try making `MutableSeq` a subclass of `Seq` (for backward compatibility if not ease of implementation).

As the mutable classes and immutable classes in Python (e.g. `list` vs `tuple`, `set` vs `frozenset`, `bytearray` vs `bytes`) are not subclasses of each other, I think we should follow the same pattern for `Seq` and `MutableSeq`.

**Comment by peterjc** [comment]
That's another good reason.

Abstract Base Classes avoids this 👍

**Pull request merged**